balanced_clustering/utils/_emi.py

```diff
@@ -2,32 +2,17 @@
 #           Corey Lynch <coreylynch9@gmail.com>
 # License: BSD 3 clause
 
-from libc.math cimport exp, lgamma
-from scipy.special import gammaln
 import numpy as np
-cimport numpy as np
-cimport cython
+import numba
+from scipy.sparse import spmatrix
+from math import exp, lgamma
 
-np.import_array()
-ctypedef np.float64_t DOUBLE
-
-def expected_mutual_information(contingency, int n_samples):
-    """Calculate the expected mutual information for two labelings."""
-    cdef int R, C
-    cdef DOUBLE N, gln_N, emi, term2, term3, gln
-    cdef np.ndarray[DOUBLE] gln_a, gln_b, gln_Na, gln_Nb, gln_nij, log_Nnij
-    cdef np.ndarray[DOUBLE] nijs, term1
-    cdef np.ndarray[DOUBLE] log_a, log_b
-    cdef np.ndarray[np.int32_t] a, b
-    #cdef np.ndarray[int, ndim=2] start, end
-    R, C = contingency.shape
-    N = <DOUBLE>n_samples
-    a = np.ravel(contingency.sum(axis=1).astype(np.int32, copy=False))
-    b = np.ravel(contingency.sum(axis=0).astype(np.int32, copy=False))
+@numba.njit(fastmath=True, cache=True, parallel=True)
+def _emi(a, b, R, C, N):
     # There are three major terms to the EMI equation, which are multiplied to
     # and then summed over varying nij values.
     # While nijs[0] will never be used, having it simplifies the indexing.
-    nijs = np.arange(0, max(np.max(a), np.max(b)) + 1, dtype='float')
+    nijs = np.arange(0.0, max(np.max(a), np.max(b)) + 1, dtype='float64')
```

Comment:
Replacing gammaln with lgamma seems to work (scipy.special isn't supported by numba). Running into typing errors for a few lines right now though (15,  33, 35).

Reply:
can you see if it works without any explicit typing?

Reply:
Line 15 is fixed with converting all inputs to float and no explicit typing: `nijs = np.arange(0.0, float(max(np.max(a), np.max(b)) + 1))`  (pushed)

Line 33 doesn't work with or without explicit typing:
```
Exception has occurred: TypingError
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<built-in function setitem>) found for signature:
 
 >>> setitem(array(undefined, 1d, C), int64, array(int64, 1d, C))
 
There are 16 candidate implementations:
   - Of which 16 did not match due to:
   Overload of function 'setitem': File: <numerous>: Line N/A.
     With argument(s): '(array(undefined, 1d, C), int64, array(int64, 1d, C))':
    No match.
```

Haven't quite figured out what this corresponds to yet, because both `a` and `b` here have declared types.